File: dsa/5.binary_search/540_Single_Element_in_a_Sorted_Array.java
package binary_search;

class Solution {
    public int singleNonDuplicate(int[] nums) {
        int left = 0;
        int right = nums.length - 1;

        while (left <= right){
            int mid = left + (right-left)/2;
            // System.out.printf("mid: %d, left: %d, right:%d\n", mid, left, right);
            int rightVal = -1;
            int leftVal = -1;
            
            if (mid + 1 < nums.length){
                rightVal = nums[mid+1];
            }

            if (mid - 1 >= 0){
                leftVal = nums[mid-1];
            }

            if (nums[mid] != leftVal & nums[mid] != rightVal){
                return nums[mid];
            }
            else {
                if ((mid % 2 == 0 & nums[mid] == rightVal & nums[mid] != leftVal) 
                    || (mid % 2 == 1 & nums[mid] == leftVal & nums[mid] != rightVal)){
                        left = mid + 1;
                    }
                else {
                    right = mid - 1;
                }
            }

        }
        return -1;
    }
}

File: dsa/5.binary_search/540_Single_Element_in_a_Sorted_Array.py
import typing as t

class Solution:
    def singleNonDuplicate(self, nums: t.List[int]) -> int:
        left: int = 0
        right: int = len(nums) - 1

        while left <= right:
            mid: int = int((right - left) / 2) + left
            left_val: int = nums[mid - 1] if mid - 1 >= 0 else -1
            right_val: int = nums[mid + 1] if mid + 1 < len(nums) else -1

            if nums[mid] != left_val and nums[mid] != right_val:
                return nums[mid]

            if(
                (mid % 2 == 0 and nums[mid] == left_val)
                or
                (mid % 2 == 1 and nums[mid] == right_val)
            ):
                right = mid - 1
            else:
                left = mid + 1
        
        return -1 